### ISSUE TITLE
Simplest case of miscompiling compound conditional

### ISSUE THREAD

**Issue by KeenS** [mirah/mirah#285]
(regarding google group discussion) I found this code being miscompiled.

```mirah
if true or nil or false
  puts "OK"
else
  puts "NG"
end
```

It passes compilation, but raise an exception when run.

```
Exception in thread "main" java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    If.main([Ljava/lang/String;)V @21: ifeq
  Reason:
    Type top (current frame, stack[0]) is not assignable to integer
  Current Frame:
    bci: @21
    flags: { }
    locals: { '[Ljava/lang/String;', integer }
    stack: { top }
  Bytecode:
    0x0000000: 043c 1b99 0007 1ba7 000e 014d 2cc6 0007
    0x0000010: 2ca7 0004 0399 000e b200 1312 15b6 001b
    0x0000020: a700 0bb2 0013 121d b600 1bb1          
  Stackmap Table:
    append_frame(@10,Integer)
    append_frame(@20,Null)
    full_frame(@21,{Object[#36],Integer},{Top})
    same_frame(@35)
    same_frame(@43)

	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2693)
	at java.lang.Class.privateGetMethodRecursive(Class.java:3040)
	at java.lang.Class.getMethod0(Class.java:3010)
	at java.lang.Class.getMethod(Class.java:1776)
	at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:544)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:526)
```

It seems a bit different from the error case discussed at google group.

**Comment by KeenS** [comment]
I tested some more cases

* `||` instead of `or`
error

* `nil` instead of `false`
error

* `false` instead of `nil`
*no error*

* all are `true`
*no error*

* all are `nil`
*no error*

* all are `false`
*no error*

Then, it seems `nil` after `true` is suspicious. But

* `true || nil`
*no error*

* `true || true || nil`
*no error*

* `true || nil && nil`
*no error*

* `true or nil || nil`
*no error*

* `true || nil or nil`
error

To sum up,

```
        or
      /     \
true      or
           /    \
        nil      *
```

would be enbuged.

**Comment by Jscottkasten** [comment]
Here's my original code from the discussion thread.  It is roughly comparable to the comments above, except it is nil?, nil?, boolean.  Hopefully this provides enough examples and some "unit test" code to get it resolved.

package org.mirah.unit.test.ComplexConditional

#
# Std library support.
# 
import java.lang.Object
import java.lang.String
import java.util.ArrayList
import java.util.HashMap




#####
#
# Demonstration type one.
#
class TypeOne
  attr_reader val:int

  def initialize
    @val = 1
  end
end

#####
#
# Demonstration type two.
#
class TypeTwo
  attr_reader val:int

  def initialize
    @val = 2
  end
end

#####
#
# Demonstration type three.
#
class TypeThree
  attr_reader str:String

  def initialize (str:String)
    @str = str
  end
end

#####
#
# Demonstration class for compiler issue.
#
class ComplexConditional
  def initialize
    @type1 = TypeOne(nil)  # For type inference.
    @type2 = TypeTwo(nil)  # For type inference.
    @type3 = TypeThree.new("Test Me!")
  end

  def three_bool_if:boolean
    if @type1 or @type2 or @type3.str.equals("__EMPTY__")
      # If this line is called, the test FAILED.
      result = false
    else
      # If this line is called, the test PASSED.
      result = true
    end

    return result
  end
end

**Comment by Jscottkasten** [comment]
This is my sample code from the initial discussion. It is similar to the above in that it follows the pattern nil?, nil?, boolean.  This should help nail down the issue and provide a "unit test" case.

In addition to my comments from before, I found that even this form was executing incorrectly:

```
if type1 or type2  # nil? or nil?
elsif bool_func()
else
end
```

The final "else" was getting executed when it should not have.

```

package org.mirah.unit.test.ComplexConditional

#
# Std library support.
# 
import java.lang.Object
import java.lang.String
import java.util.ArrayList
import java.util.HashMap




#####
#
# Demonstration type one.
#
class TypeOne
  attr_reader val:int

  def initialize
    @val = 1
  end
end

#####
#
# Demonstration type two.
#
class TypeTwo
  attr_reader val:int

  def initialize
    @val = 2
  end
end

#####
#
# Demonstration type three.
#
class TypeThree
  attr_reader str:String

  def initialize (str:String)
    @str = str
  end
end

#####
#
# Demonstration class for compiler issue.
#
class ComplexConditional
  def initialize
    @type1 = TypeOne(nil)  # For type inference.
    @type2 = TypeTwo(nil)  # For type inference.
    @type3 = TypeThree.new("Test Me!")
  end

  def three_bool_if:boolean
    if @type1 or @type2 or @type3.str.equals("__EMPTY__")
      # If this line is called, the test FAILED.
      result = false
    else
      # If this line is called, the test PASSED.
      result = true
    end

    return result
  end
end
```
